egs/wsj/s5/steps/cleanup/internal/align_ctm_ref.py

```diff
@@ -0,0 +1,213 @@
+import sys, argparse, imp
+
+verbosity = 0
+
+def GetArgs():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("--hyp-format", type=str, choices = ["Text", "CTM"],
+                        "Format used for the hypothesis")
+    parser.add_argument("--eps-symbol", type=str, default = "-",
+                        "Symbol used to contain alignment to empty symbol")
+    parser.add_argument("--verbose", type=int, default=0)
+    parser.add_argument("--correct-score", type=int, default=1,
+                        "Score for correct matches")
+    parser.add_argument("--substitution-penalty", type=int, default=1,
+                        "Penalty for substitution errors")
+    parser.add_argument("--deletion-penalty-type", type=str, default = "linear",
+                        choices = ["linear"],
+                        help = "Type of function for deletion penalty")
+    parser.add_argument("--deletion-penalty-scale", type-int, default=1,
+                        "Scale on deletion penalty")
+    parser.add_argument("--insertion-penalty-type", type=str, default = "linear",
+                        choices = ["linear"],
+                        help = "Type of function for insertion penalty")
+    parser.add_argument("--insertion-penalty-scale", type-int, default=1,
+                        "Scale on insertion penalty")
+    parser.add_argument("--debug-only", type=str, default = "false",
+                        choices = ["true", "false"],
+                        help = "Run test functions only")
+    parser.add_argument("ref", type=str,
+                        help = "Reference text")
+    parser.add_argument("hyp", type=str,
+                        help = "Hypothesis text or CTM")
+
+    args = parser.parse_args()
+    verbosity = args.verbose
+    args.debug_only = True if args.debug_only == "true" else False
+
+    return args
+
+def SmithWatermanAlignment(ref, hyp, eps_symbol,
+                           SimilarityScore, DelScore, InsScore,
+                           hyp_format = "CTM"):
+    output = []
+
+    M = len(ref)
+    N = len(hyp)
+
+    bp = [ [ ] for x in range(M+1) ]
+    H = [ [ ] for x in range(M+1) ]
+
+    for m in range(M+1):
+        H[m] = [ 0 for x in range(N+1) ]
+        bp[m] = [ (0,0) for x in range(N+1) ]
+
+    global_max = 0
+    global_max_element = (0,0)
+
+    for m in range(1, M+1):
+        for n in range(1, N+1):
+            sub_or_ok = (H[m-1][n-1] +
+                         SimilarityScore(ref[m-1],
+                                         hyp[n-1][3] if hyp_format == "CTM"
+                                                   else hyp[n-1]))
+            if sub_or_ok > 0:
+                H[m][n] = sub_or_ok
+                bp[m][n] = (m-1, n-1)
+                if verbosity > 2:
+                    print ("({0},{1}) -> ({2},{3}): {4} ({5},{6})".format(m-1,n-1,m,n, H[m][n], ref[m-1], hyp[n-1]), file = sys.stderr)
+
+            for k in range(1, m+1):
```

Comment:
This algorithm looks like it is cubic in the length of the sequences.  That doesn't strike me as being very practical-- at least, not if we are talking about long texts.  E.g. imagine if the text is the Qur'an, that was one of the potential applications.   Cubic in the size of that is too slow.
Now, let's assume that the num-words in the hypothesis is reasonably small (because we're limiting ourselves to 30 second or 1 minute segments of text).  You'll need to find a practical method to find candidate segments of reasonable length within the reference text.
Now, suppose we're doing batch processing of a bunch of these segments, so we can afford to preprocess the reference text somehow.  I suggest to do the following: segment the reference text into "documents" of fixed length, e.g. 1000 words, and for each document compute n-gram stats up to trigram.  For each n-gram, you compute its overall n-gram count, and its count for each "document" that it appears in.
Then, for the decoded segment you're trying to align, compute counts of n-grams up to the same order.  Compute a TF-IDF based similarity of the segment with each of the documents, based on the n-gram counts.  You can use that to select candidate documents-  e.g. take the document with the highest TF-IDF, and eat up documents adjacent to it that have high TF-IDF, or some modification of that.  Of course, what this gives you is a sub-range of the entire text; the notion of "document" is meaningless here, and if exported to any of our external interfaces it would add extra un-needed complexity to the system.

What I think is a workable approach is that each Smith-Waterman alignment process gives you, for a particular 30 second to 1 minute segment of ctm (these are actually stitched-together ctm's)... gives you a range [start-word, end-word] of the reference text that "aligns best".  You might want to just reject the entire segment if the alignment score was below some threshold (e.g. 0.75 per word or something).
You can then use the program "align-text" to align this "best reference" to the text from the ctm, making the ctm-edits format, and put it through some of the same stages used in clean_and_segment_data.sh, i.e. modify_ctm_edits.py, taint_ctm_edits.py, segment_ctm_edits.py.  This will only keep any text that aligns perfectly (or almost perfectly) within the shorter sub-segments.